### PR TITLE
Fix a race involving snapshots and the right-hand side of a split.

### LINKED ISSUE #1191: crash in raft.sendAppend: requested index at entry is unavailable
This happened in tandem with #1185 and after a flurry of messages

```
W0529 13:18:33.251543       1 dist_sender.go:588] failed to invoke InternalPushTxn: key range "00065201
"-"" outside of bounds of range ""-""
```

(which are not surprising in this scenario given that the node never executes any of the splits)

```
I0529 13:18:33.251657       1 retry.go:95] routing InternalPushTxn rpc failed; retrying immediately
W0529 13:18:33.251948       1 dist_sender.go:588] failed to invoke InternalPushTxn: key range "00065201"-"" outside of bounds of range ""-""
I0529 13:18:33.251966       1 retry.go:95] routing InternalPushTxn rpc failed; retrying immediately
W0529 13:18:33.253062       1 dist_sender.go:588] failed to invoke InternalPushTxn: key range "00065201"-"" outside of bounds of range ""-""
I0529 13:18:33.253138       1 retry.go:95] routing InternalPushTxn rpc failed; retrying immediately
panic: requested entry at index is unavailable

goroutine 64 [running]:
github.com/coreos/etcd/raft.(*raft).sendAppend(0xc2097f4000, 0x300000003)
        /home/tobias/Code/Go/src/github.com/coreos/etcd/raft/raft.go:250 +0x159
github.com/coreos/etcd/raft.stepLeader(0xc2097f4000, 0x4, 0x200000002, 0x300000003, 0x6, 0x0, 0xa, 0x0, 0x0, 0x0, ...)
        /home/tobias/Code/Go/src/github.com/coreos/etcd/raft/raft.go:523 +0x6f3
github.com/coreos/etcd/raft.(*raft).Step(0xc2097f4000, 0x4, 0x200000002, 0x300000003, 0x6, 0x0, 0xa, 0x0, 0x0, 0x0, ...)
        /home/tobias/Code/Go/src/github.com/coreos/etcd/raft/raft.go:487 +0x259
github.com/coreos/etcd/raft.(*multiNode).run(0xc2080ff0e0)
        /home/tobias/Code/Go/src/github.com/coreos/etcd/raft/multinode.go:236 +0xd0b
created by github.com/coreos/etcd/raft.StartMultiNode
        /home/tobias/Code/Go/src/github.com/coreos/etcd/raft/multinode.go:56 +0xb6
```

### LINKED ISSUE #1242: Manual range split causes panic
Starting a 3 node cluster on AWS and ran the bank example against it. No problems.
Added another 2 nodes.

Attempted a range split:
```
./cockroach range split 000000500 --addr=cockroach-db-1441359306.us-east-1.elb.amazonaws.com:8080 --insecure
```

Two nodes died.

The first one with:
```
W0603 14:11:51.065761      11 store.go:1179] failed to resolve on inconsistent read: replica 1:1 not leader; leader is 3:3
W0603 14:11:51.065969      11 store.go:1179] failed to resolve on inconsistent read: replica 1:1 not leader; leader is 3:3
panic: intents:<key:"\000\000\000k\000\001rdsc" txn:<name:"split range 1 at \"000000012\"" key:"000000012" id:"\021l\377R\345\307O\330\255\202\260\016\177\274\365)" priority:2037030288 isolation:SERIALIZABLE status:PENDING epoch:0 timestamp:<wall_time:1433340701486318134 logical:0 > orig_timestamp:<wall_time:1433340701486318134 logical:0 > max_timestamp:<wall_time:1433340701736318134 logical:0 > certain_nodes:<> > > resolved:false
```

The other node died with:
```
E0603 14:11:55.751688      11 raft_transport.go:174] raft message to node 1 failed: read tcp 172.31.22.174:8080: connection reset by peer
I0603 14:11:55.770180      11 client.go:233] client 172.31.22.174:8080 heartbeat failed: read tcp 172.31.22.174:8080: connection reset by peer; recycling...
I0603 14:11:55.770343      11 client.go:233] client 172.31.22.174:8080 heartbeat failed: read tcp 172.31.22.174:8080: connection reset by peer; recycling...
W0603 14:11:55.771377      11 raft_transport.go:170] raft client for node 1 closed
I0603 14:11:55.771523      11 gossip.go:460] client disconnected: read tcp 172.31.22.174:8080: connection reset by peer
panic: requested entry at index is unavailable
```

### PR DESCRIPTION
**Pull request by bdarnell**
After a split, the newly-created range has multiple members (same as the
parent range) and can therefore complete elections and send raft
traffic, but the log index is reset. Since the range is considered
initialized, we need to reset the appliedIndex to raftInitialLogIndex
instead of 0 (which signifies an uninitialized range). If a range became
leader with an appliedIndex of 0, it could panic while trying to
generate a snapshot to send to a lagging follower.

Commit dd45ddda fixed a related issue involving the left-hand side of a
split.

Fixes #1191.
Fixes #1242.

### CODE REVIEW

**Comment by spencerkimball** [comment]
LGTM

File: storage/range_raftstorage.go [hunk 1]
```diff
@@ -169,7 +169,12 @@ func (r *Range) FirstIndex() (uint64, error) {
 
 // loadAppliedIndex retrieves the applied index from the supplied engine.
 func (r *Range) loadAppliedIndex(eng engine.Engine) (uint64, error) {
-	appliedIndex := uint64(0)
+	var appliedIndex uint64
+	if r.isInitialized() {
+		appliedIndex = raftInitialLogIndex
+	} else {
+		appliedIndex = 0
```
Comment (tschottdorf) [inline storage/range_raftstorage.go]: The `else` branch is unnecessary.

**Comment by tschottdorf** [comment]
LGTM, though I can't find `dd45ddda` from your description.

**Comment by tschottdorf** [comment]
btw, `TestRangeDescriptorSnapshotRace` still takes 3-4 seconds. We might want to go from 40 to a total of 10 retries once we're sure that the race is gone for good.

**Pull request merged**